feature-pack-api/src/test/java/org/jboss/provisioning/state/xml/test/ProvisionedStateXmlWriterTestCase.java

```diff
@@ -0,0 +1,237 @@
+                                .setIdParam("id-param", "a")
+                                .build()
+                        )
+                        .addFeature(ProvisionedFeatureBuilder.builder(
+                                ResolvedFeatureId.create(ArtifactCoords.newGav("org.jboss.group", "fp1", "1.0"),
+                                        "spec1", "create-param", "c"))
+                                .build())
+                        .build())
+                .build();
+
+        // marshall and unmarshall
+        Path path = marshallToTempFile(originalState);
+        ProvisionedState newState = validator.validateAndParse(path);
+
+        // read features and check the order
+        ReadFeaturesHandler handler = new ReadFeaturesHandler();
+        newState.getConfigs().iterator().next().handle(handler);
+        Iterator<ProvisionedFeature> iterator = handler.features.iterator();
+
+        Assert.assertEquals("org.jboss.group:fp1:1.0#spec1:create-param=a", iterator.next().getId().toString());
+        Assert.assertEquals("org.jboss.group:fp1:1.0#spec2:create-param=b", iterator.next().getId().toString());
+        Assert.assertEquals("org.jboss.group:fp1:1.0#spec1:create-param=c", iterator.next().getId().toString());
+        Assert.assertEquals("org.jboss.group:fp2:1.0#spec3:id-param=b", iterator.next().getId().toString());
+        Assert.assertEquals("org.jboss.group:fp2:1.0#spec3:id-param=a", iterator.next().getId().toString());
+        Assert.assertEquals("org.jboss.group:fp1:1.0#spec1:create-param=c", iterator.next().getId().toString());
+    }
+
+    @Test
+    public void testParams() throws Exception {
+        ProvisionedState originalState = ProvisionedState.builder()
+                .addConfig(ProvisionedConfigBuilder.builder()
+                        .addFeature(ProvisionedFeatureBuilder.builder(
+                                new ResolvedSpecId(ArtifactCoords.newGav("org.jboss.group", "fp2", "1.0"), "spec3"))
+                                .setIdParam("id-param", "a")
+                                .setConfigParam("config-param", "value")
+                                .setParam("param", "value")
+                                .setResolvedParam("resolved-param", "value")
+                                .build())
+                        .build())
+                .build();
+
+        // marshall and unmarshall
+        Path path = marshallToTempFile(originalState);
+        ProvisionedState newState = validator.validateAndParse(path);
+
+        // retrieve the feature
+        ReadFeaturesHandler readFeaturesHandler = new ReadFeaturesHandler();
+        newState.getConfigs().get(0).handle(readFeaturesHandler);
+        ProvisionedFeature feature = readFeaturesHandler.features.get(0);
+
+        // resolved params should contain params added by setParam(), setResolvedParam() and setIdParam()
```

Comment:
This is to be verified - I suppose only resolved params should be marshalled and then read again as resolved params.

Reply:
Resolved and config are two representations of the same parameter value. The config one is the string value which is read from the xml. The resolved one (which is parsed from the config) is used at runtime by the provisioning mechanism but when the provisioned state is persisted we have to store the parameters as strings again.

Reply:
Ah, it started to make sense now :).

Another two things I want to check:

1) when I set an IdParam, it's marshalled twice, once as a <param> element where the value now have it's config representation, and once inside a feature id string, where there is #toString() of it's resolved representation:

```
                <feature id="org.jboss.group:fp:1.0#spec:id-param=resolved">
                    <param name="id-param" value="config"/>
                </feature>
```

Should the ID string contain also the config representation of the parameter?

2) When XML is parsed by current parser, then ProvisionedFeatureBuilder#getParamNames() returns only params names from the resolvedParams map (which are only id params, normal params are missing since they are only in configParams map). There is no other way to read param names that are only in configParams map, is that fine?

Also, getResolvedParam(String name) throws unsupported exception while getResolvedParams() is implemented.

ProvisionedFeatureBuilder:
```
    ...
    @Override
    public boolean hasParams() {
        return !resolvedParams.isEmpty();
    }

    @Override
    public Collection<String> getParamNames() {
        return resolvedParams.keySet();
    }

    @Override
    public Object getResolvedParam(String name) {
        throw new UnsupportedOperationException();
    }

    @Override
    public String getConfigParam(String name) throws ProvisioningException {
        return configParams.get(name);
    }

    @Override
    public Map<String, Object> getResolvedParams() {
        return resolvedParams;
    }
    ...
```

Reply:
1. Yes, that's a good catch.
2. I think we should drop the resolved params completely in this class. At least for now. We simply can't resolve parameter values w/o a few other classes from the provisioning runtime. That means we can't also create proper feature IDs (unless the id param values are strings).

Thanks!